src/backend/distributed/commands/alter_table.c

```diff
@@ -360,6 +360,54 @@ worker_change_sequence_dependency(PG_FUNCTION_ARGS)
 }
 
 
+/*
+ * UndistributeTables undistributes given relations. It first collects all foreign keys
+ * to recreate them after the undistribution. Then, drops the foreign keys and
+ * undistributes the relations. Finally, it recreates foreign keys.
+ */
+void
+UndistributeTables(List *relationIds)
+{
+	List *tablesToConvert = NIL;
+	Oid relationId = InvalidOid;
+	foreach_oid(relationId, relationIds)
+	{
+		/*
+		 * Skip partitions as they would be distributed/undistibuted by partitioned table.
+		 *
+		 * We should filter out partitions here before calling CreateTenantSchemaTable.
+		 * Otherwise, converted partitioned table would change oid of partitions and its
+		 * partition tables would fail wih oid not exist.
+		 */
+		if (PartitionTable(relationId))
+		{
+			continue;
+		}
+		tablesToConvert = lappend_oid(tablesToConvert, relationId);
+	}
+
+	/*
+	 * Collect foreign keys for recreation and then drop fkeys and undistribute
+	 * tables.
+	 */
+	List *originalForeignKeyRecreationCommands = NIL;
+	foreach_oid(relationId, tablesToConvert)
+	{
+		List *fkeyCommandsForRelation =
+			GetFKeyCreationCommandsRelationInvolvedWithTableType(relationId,
+																 INCLUDE_ALL_TABLE_TYPES);
+		originalForeignKeyRecreationCommands = list_concat(
+			originalForeignKeyRecreationCommands, fkeyCommandsForRelation);
+		DropFKeysAndUndistributeTable(relationId);
```

Comment:
We should probably implement a multi-undistribute at some point, because after the first table we block reads on that table for a while.

Reply:
no, I mean that we do all the data movement first, and only drop and replace the old tables until the very end

now we do data movement and drop and replace for each table one by one.

Reply:
(not for this PR)